### PR TITLE
Added new metric type Event.

### PR DESCRIPTION
**Pull request by GitAtSmsPortal**
Added cleaner to clean out events after a period of time after they have been reported.
Updated the MetricTags GetHashCode function to be public static so that we can get the hash for a strin[] of tags in multilpe places.
Updated the reportes to report the event metrics.

### CODE REVIEW

**Review by wauk — APPROVED** [review]

File: Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+using Xunit;
+
+namespace Metrics.Tests.Reporting
+{
+    public class EventMetricsCleanerTest
+    {
+        [Fact]
+        public void StartsWithZeroReports()
+        {
+            EventMetricsCleaner.Clear();
```
Comment (Francois-du-Plessis) [inline Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs]: Clear should not be called here.

File: Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+            EventMetricsCleaner.RegisterReport(new TimeSpan(0));
+
+            EventMetricsCleaner.TotalReports.Should().Be(1);
+        }
+
+        [Fact]
+        public void RegisterReport_WithGreaterTimespan_UpdatesIntervalToBeGreaterThanReportInterval()
```
Comment (Francois-du-Plessis) [inline Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs]: Could test WithSmallerTimespam_DoesNotUpdateInterval

File: Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            EventMetricsCleaner.UpdateTotalReportedEvents(reportIndex1, registry.DataProvider.Events);
+            metric.Record();
+            EventMetricsCleaner.UpdateTotalReportedEvents(reportIndex2, registry.DataProvider.Events);
+
+            EventMetricsCleaner.GetReportsReportedEventDetailCount(reportIndex1, "test").Should().Be(1);
+            EventMetricsCleaner.GetReportsReportedEventDetailCount(reportIndex2, "test").Should().Be(2);
+
+            timer.OnTimerCallback();
+
+            EventMetricsCleaner.GetEventDetailCount("test").Should().Be(1);
```
Comment (Francois-du-Plessis) [inline Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs]: Should this not rather look at the event list in the DefaultMetricsRegistry to see if it actually clears the event

File: Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs (hunk 2 above)
Comment (GitAtSmsPortal) [reply Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs]: Have added a unit test to test the correct logic.

File: Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs (hunk 1 above)
Comment (GitAtSmsPortal) [reply Src/Metrics.Tests/Reporting/EventMetricsCleanerTest.cs]: Right you are. Have made all other tests clear after they have run.

**Review by Francois-du-Plessis — APPROVED** [review]

**Pull request merged**